cardano-node/src/Cardano/Node/Handlers/TopLevel.hs

```diff
@@ -74,9 +75,16 @@ toplevelExceptionHandler prog = do
       ]
   where
     -- Let async exceptions rise to the top for the default GHC top-handler.
-    -- This includes things like CTRL-C.
+    -- This includes things like CTRL-C. If a linked thread throws ExitSuccess,
+    -- then we rethrow ExitSuccess. This happens for example when using the
+    -- `--shutdown-on-slot-synced` option.
     rethrowAsyncExceptions :: SomeAsyncException -> IO a
-    rethrowAsyncExceptions = throwIO
+    rethrowAsyncExceptions full@(SomeAsyncException e) = do
+      case fromException (toException e) of
+        Just (ExceptionInLinkedThread _ eInner)
+          | Just ExitSuccess <- fromException eInner
+          -> throwIO ExitSuccess
+        _ -> throwIO full
```

Comment:
How does propagating `ExitSuccess` this way change the observed behaviour of the program?

Reply:
I've added info in the PR description. See point 2.